orc8r/cloud/go/obsidian/access/rest_middleware.go

```diff
@@ -100,3 +104,104 @@ func makeErr(decorate logDecorator, status int, errFmt string, errArgs ...interf
 	glog.V(1).Infof("REST middleware (obsidian) rejected request: %s", decorate(errFmt, errArgs...))
 	return echo.NewHTTPError(status, fmt.Sprintf(errFmt, errArgs...))
 }
+
+// TokenMiddleware parses the <username>:<token> from the header, validates the token,
+// then checks if the request is within the specified permissions granted to the user
+func TokenMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		req := c.Request()
+		header := req.Header.Get(CLIENT_ACCESS_TOKEN_KEY)
+
+		username, token, err := parseAuthHeader(header)
+		if err != nil {
+			// TODO(christinewang5): remove after bootstrapping admin token
+			glog.Error("no header found, bootstrapping admin user")
+			adminToken, err := bootstrapAdmin(req.Context())
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, err)
+			}
+			username = RootUsername
+			token = adminToken
+			// return echo.NewHTTPError(http.StatusBadRequest, err)
+		}
+
+		if err := certifier.ValidateToken(token); err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, err)
+		}
+
+		// make sure that token is registered with user
+		getOpReq := certifierprotos.GetUserRequest{
+			Username: username,
+			Token:    token,
+		}
+		tokensList, err := certifier.GetUserTokens(req.Context(), &getOpReq)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, err)
+		}
+
+		// take tokenList, request type, resource and exchange for permission decision
+		requestType := getRequestAction(req, nil)
+		resource := req.RequestURI
+		getPDReq := certifierprotos.GetPolicyDecisionRequest{
+			TokenList:     tokensList,
+			RequestAction: requestType,
+			Resource:      resource,
+		}
+		pd, err := certifier.GetPolicyDecision(req.Context(), &getPDReq)
+		if err != nil || pd.Effect == certifierprotos.Effect_DENY {
+			return echo.NewHTTPError(http.StatusForbidden, err)
+		}
```

Comment:
Does echo.NewHTTPError not already log things when it is processed through the middleware (from one of my first bootcamp tasks lol)? It doesn't log the stack trace though, should I just double log it?
https://github.com/christinewang5/magma/blob/auth_token/orc8r/cloud/go/service/service.go#L177-L192

Reply:
Ah right so -- the distinction here is, generally speaking, we want the "client-facing" part of an internal error to be security-considerate, meaning we don't give "all" the info away e.g. stack trace

We don't do a great job of being consistent about that. But the idea is you give a basic "client-facing" err, and log the details of the err e.g. stack trace

Is not end of the world though, up to you since we're not consistent about it